### PR TITLE
Updated the scheduler so that it can run in synchronous or asynchronous ...

### PR DESCRIPTION
**Pull request by nevermindewe**
...mode.

In synchronous mode, only one thread in the pool will be used.
This is not a complete/fully baked solution; it is incremental progress.
Things that need to be done before it can be considered fully baked:
 - deciding whether or not to create the threadpool in synchronous mode
 - adding a configuration mechanism to specify sync/async mode when threaded.get is called. (Currently it relies on kwargs to determine sync or async)

To support sync mode, the "state" dict locks had to be rearranged.
Now, locking is a little finer-grained and immediately surrounds the code that updates the state dict.

Side-effects of sync mode:
 - state['num-active-threads'] will never be non-zero (based on when it is updated).
   This means that the queue would never be emptied while jobs were waiting to run because "state['num-active-threads'] < nthreads" was always true.
   This was solved by adding a max_queue_size variable that tells the maximum number of items allowed in the queue before we start emptying it.

### CODE REVIEW

File: dask/threaded.py [hunk 1]
```diff
@@ -242,40 +241,41 @@ def execute_task(dsk, key, state, queue, results, lock):
...
-        s = state['waiting'][dep]
-        s.remove(key)
-        if not s:
-            del state['waiting'][dep]
-            state['ready'].add(dep)
-
-    for dep in state['dependencies'][key]:
-        if dep in state['waiting_data']:
-            s = state['waiting_data'][dep]
+    with lock:
```
Comment (mrocklin) [inline dask/threaded.py]: One of the reasons I didn't put the lock into `finish_task` was that I wanted it to be easier to test.  Ideally we can isolate heavy state manipulation logic from the secondary logic of locking.  We should only merge the two if locking logic needs to interleave with state-manipulation logic (as is the case with fine-grained locking).

In this case it doesn't seem like `finish_task` needs to control the lock (nothing here is locked in a fine grained way).

File: dask/threaded.py [hunk 2]
```diff
@@ -242,40 +241,41 @@ def execute_task(dsk, key, state, queue, results, lock):
...
+            elif dep in state['cache'] and dep not in results:
                 release_data(dep, state)
-                assert dep not in state['cache']
-        elif dep in state['cache'] and dep not in results:
-            release_data(dep, state)
 
-    state['finished'].add(key)
-    state['num-active-threads'] -= 1
+        state['finished'].add(key)
+        state['num-active-threads'] -= 1
```
Comment (mrocklin) [inline dask/threaded.py]: I've merged  `state['num-active-threads']` and `jobs` to a single set of running tasks `state['running']` in #8 .  I'm happy to handle the merge conflicts in this case.

**Comment by nevermindewe** [comment]
@mrocklin I merged your updates with the sync/async stuff. All the tests still pass, and I think it is still correct.

**Comment by mrocklin** [comment]
Closing in favor of #33

**Pull request closed without merging**